Compilation error in Vaadin 24

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.1:testCompile (default-testCompile) on project year-month-calendar: Compilation failure
[ERROR] /J:/fc/addons/YearMonthCalendarAddon/src/test/java/com/flowingcode/addons/ycalendar/YearDemo.java:[61,14] cannot find symbol
[ERROR]   symbol:   method setHasControls(boolean)
[ERROR]   location: variable yearField of type com.vaadin.flow.component.textfield.IntegerField
```

https://github.com/FlowingCode/YearMonthCalendarAddon/blob/0d84347b9a44dd28b710803a394d0879cbd7692f/src/test/java/com/flowingcode/addons/ycalendar/YearDemo.java#L61

Note that `setHasControls` has been deprecated in 23.3 (in favor of `setStepButtonsVisible(boolean)`) and it was removed in Vaadin 24. The component and the demo needs to be compatible with Vaadin 22, 23.2, 23.3 and 24.

--- a/src/test/java/com/flowingcode/addons/ycalendar/TestUtils.java
+++ b/src/test/java/com/flowingcode/addons/ycalendar/TestUtils.java
@@ -19,10 +19,12 @@
  */
 package com.flowingcode.addons.ycalendar;
 
+import java.lang.reflect.Method;
 import java.time.LocalDate;
 import java.util.Arrays;
 import java.util.HashSet;
 import java.util.List;
+import java.util.Optional;
 import java.util.Set;
 
 public class TestUtils {
@@ -92,4 +94,12 @@ public class TestUtils {
     }
   }
 
+  public static Optional<Method> getMethod(Class<?> clazz, String methodName, Class<?>... args) {
+    try {
+      return Optional.ofNullable(clazz.getMethod(methodName, args));
+    } catch (NoSuchMethodException e) {
+      return Optional.empty();
+    }
+  }
+  
 }

--- a/src/test/java/com/flowingcode/addons/ycalendar/YearDemo.java
+++ b/src/test/java/com/flowingcode/addons/ycalendar/YearDemo.java
@@ -28,6 +28,7 @@ import com.vaadin.flow.component.orderedlayout.HorizontalLayout;
 import com.vaadin.flow.component.textfield.IntegerField;
 import com.vaadin.flow.router.PageTitle;
 import com.vaadin.flow.router.Route;
+import java.lang.reflect.Method;
 import java.time.DayOfWeek;
 
 @DemoSource
@@ -35,6 +36,11 @@ import java.time.DayOfWeek;
 @CssImport(value = "./styles/test_year-month-calendar.css", themeFor = "vaadin-month-calendar")
 @Route(value = "year-month-calendar/year", layout = YearMonthCalendarDemoView.class)
 public class YearDemo extends Div {
+
+  private static final Method setHasControlsMethod =
+      TestUtils.getMethod(IntegerField.class, "setHasControls", boolean.class).orElse(null);
+  private static final Method setStepButtonsVisibleMethod =
+      TestUtils.getMethod(IntegerField.class, "setStepButtonsVisible", boolean.class).orElse(null);
 
   public YearDemo() {
     YearCalendar calendar = new YearCalendar();
@@ -58,7 +64,18 @@ public class YearDemo extends Div {
     Span instructions = new Span("Use arrow keys or Ctrl+arrow keys to move.");
 
     IntegerField yearField = new IntegerField();
-    yearField.setHasControls(true);
+    try {
+      if (setHasControlsMethod != null) {
+        // Vaadin 14/22/23
+        setHasControlsMethod.invoke(yearField, true);
+      } else if (setStepButtonsVisibleMethod != null) {
+        // Vaadin 24
+        setStepButtonsVisibleMethod.invoke(yearField, true);
+      }
+    } catch (Exception ex) {
+      ex.printStackTrace();
+    }
+
     yearField.setValue(calendar.getYear());
     yearField.addValueChangeListener(e -> calendar.setYear(e.getValue()));
 

